frontend/javascripts/oxalis/view/right-menu/advanced_search_popover.js

```diff
@@ -0,0 +1,160 @@
+// @flow
+import { Icon, Input, Tooltip, Popover } from "antd";
+import * as React from "react";
+import memoizeOne from "memoize-one";
+import ButtonComponent from "oxalis/view/components/button_component";
+
+import Shortcut from "libs/shortcut_component";
+import DomVisibilityObserver from "oxalis/view/components/dom_visibility_observer";
+
+const InputGroup = Input.Group;
+
+type Props<S> = {
+  data: Array<S>,
+  searchKey: $Keys<S>,
+  onSelect: S => void,
+  children: React.Node,
+  provideShortcut?: boolean,
+};
+
+type State = {
+  isVisible: boolean,
+  searchQuery: string,
+  currentPosition: ?number,
+};
+
+export default class AdvancedSearchPopover<S: Object> extends React.PureComponent<Props<S>, State> {
+  state = {
+    isVisible: false,
+    searchQuery: "",
+    currentPosition: null,
+  };
+
+  availableOptions: Array<S> = [];
+
+  selectNextOptionWithOffset = (offset: number) => {
+    let { currentPosition } = this.state;
+    const numberOfAvailableOptions = this.availableOptions.length;
+    if (numberOfAvailableOptions === 0) {
+      return;
+    }
+    if (currentPosition == null) {
+      // If there was no previous currentPosition for the current search query,
+      // set currentPosition to an inital value.
+      currentPosition = offset >= 0 ? -1 : numberOfAvailableOptions;
+    }
+    currentPosition = (currentPosition + offset) % numberOfAvailableOptions;
+    if (currentPosition < 0) {
+      currentPosition = numberOfAvailableOptions + currentPosition;
+    }
+    this.setState({ currentPosition });
+    this.props.onSelect(this.availableOptions[currentPosition]);
+  };
+
+  selectNextOption = () => {
+    this.selectNextOptionWithOffset(1);
+  };
+
+  selectPreviousOption = () => {
+    this.selectNextOptionWithOffset(-1);
+  };
+
+  getAvailableOptionsFrom = memoizeOne(
+    (data: Array<S>, searchQuery: string, searchKey: $Keys<S>): Array<S> =>
+      searchQuery !== ""
+        ? data.filter(
+            datum => datum[searchKey].toLowerCase().indexOf(searchQuery.toLowerCase()) > -1,
+          )
+        : [],
+  );
+
+  openSearchPopover = () => {
+    this.setState({ isVisible: true });
+  };
+
+  closeSearchPopover = () => {
+    this.setState({ isVisible: false });
+  };
+
+  render() {
+    const { data, searchKey, provideShortcut, children } = this.props;
+    const { searchQuery, isVisible } = this.state;
+    let { currentPosition } = this.state;
+    currentPosition = currentPosition == null ? -1 : currentPosition;
+    this.availableOptions = this.getAvailableOptionsFrom(data, searchQuery, searchKey);
```

Comment:
A react render method should be pure and should not have side effects. Easy to fix as the memoization allows us to simply call this.getAvailableOptionsFrom everywhere we need the availableOptions, so we don't need the instance variable :)